modin/core/storage_formats/pandas/aggregations.py

```diff
@@ -0,0 +1,600 @@
+        """
+        Perform the Map phase to compute the corr/cov matrix.
+
+        In this kernel we compute all the required components to compute
+        the correlation matrix at the reduce phase, the required components are:
+            1. Matrix holding sums of pairwise multiplications between all columns
+               defined as ``M[col1, col2] = sum(col1[i] * col2[i] for i in range(col_len))``
+            2. Sum for each column (special case if there are NaN values)
+            3. Sum of squares for each column (special case if there are NaN values)
+            4. Number of values in each column (special case if there are NaN values)
+
+        Parameters
+        ----------
+        df : pandas.DataFrame
+            Partition to compute the aggregations for.
+
+        Returns
+        -------
+        pandas.DataFrame
+            A MultiIndex columned DataFrame holding the described aggregation results for this
+            specifix partition under the following keys: ``["mul", "sum", "pow2_sum", "count"]``
+        """
+        df = df.select_dtypes(include="number")
+        # It's more convenient to use a NumPy array here as it appears to perform
+        # much faster in for-loops which this kernel function has plenty of
+        raw_df = df.values.T
+        nan_mask = np.isnan(raw_df)
+        has_nans = nan_mask.sum() != 0
+
+        if has_nans:
+            if not raw_df.flags.writeable:
+                # making a copy if the buffer is read-only
+                raw_df = raw_df.copy()
+            # Replacing all NaNs with zeros so we can use much
+            # faster `np.sum()` instead of slow `np.nansum()`
+            np.putmask(raw_df, nan_mask, values=0)
+
+        cols = df.columns
+        # In this for-loop we compute a sum of pairwise multiplications between all columns
+        # result:
+        #   col1: [sum(col1 * col2), sum(col1 * col3), ... sum(col1 * colN)]
+        #   col2: [sum(col2 * col3), sum(col2 * col4), ... sum(col2 * colN)]
+        #   ...
+        sum_of_pairwise_mul = {}
+        for i, col in enumerate(cols):
+            sum_of_pairwise_mul[col] = pandas.Series(
+                np.sum(raw_df[i] * raw_df[(i + 1) :], axis=1), index=cols[i + 1 :]
+            )
+
+        # The dictionary holds an upper triangular matrix, converting it
+        # to a complete symmetrical matrix.
+        # TODO: is there a way to save space and continue with a triangular matrix?
+        sum_of_pairwise_mul = pandas.concat(sum_of_pairwise_mul, axis=1)
+        sum_of_pairwise_mul = sum_of_pairwise_mul.reindex(df.columns).reindex(
+            df.columns, axis=1
+        )
+        sum_of_pairwise_mul.update(sum_of_pairwise_mul.T)
+        np.fill_diagonal(sum_of_pairwise_mul.values, 1.0)
+
+        if has_nans:
+            sums, sums_of_squares, count = cls._compute_nan_aggs(raw_df, cols, nan_mask)
+        else:
+            sums, sums_of_squares, count = cls._compute_non_nan_aggs(df)
```

Comment:
The dimension of `sum` is different in these two cases. In one case (`_compute_non_nan_aggs`) it is a series, in the other (`_compute_nan_aggs`) it is a dataframe. Why?

Reply:
Well, it doesn't really matter what exactly we return, either a DataFrame or a Series, both of them are processed identically by `pandas.concat`

If you wish, I can manually cast those Series'ses to DataFrames

Reply:
Right, non-nan and nan cases produce different shapes of the aggregations.

For non-nan cases it's enough to compute plain sums/sums of squares/num of observations, all of them producing Series'ses:
```python
# non-nan case
>>> df
   a  b  c
0  1  3  5
1  2  3  4
2  2  4  3
3  3  5  2
# simply computing sum for each column, that's enough to compute
# proper correlation matrix
>>> df.sum() 
a     8
b    15
c    14
dtype: int64
```

In a NaN case the handling is much more complex as we need special NaN-handling logic [described here](https://github.com/dchigarev/modin/blob/035aadddcf900576fa98a8fb39b564de258e68c2/modin/core/storage_formats/pandas/aggregations.py#L259-L274):
```python
# nan-case
>>> df_nan
     a    b  c
0  1.0  NaN  5
1  2.0  3.0  4
2  NaN  4.0  3
3  3.0  5.0  2
# we have to compute pairwise sums excluding NaN values from each column individually
>>> compute_nan_aggs(df_nan)["sum"]
     a     b     c
a  6.0   8.0  11.0
b  5.0  12.0   9.0
c  6.0  12.0  14.0
```

It's not a problem if different partitions of the same frame would produce different-shaped results, as [it's possible to align them at the Reduce phase](https://github.com/dchigarev/modin/blob/035aadddcf900576fa98a8fb39b564de258e68c2/modin/core/storage_formats/pandas/aggregations.py#L402-L421)

Reply:
> It's not a problem if different partitions of the same frame would produce different-shaped results, as [it's possible to align them at the Reduce phase](https://github.com/dchigarev/modin/blob/035aadddcf900576fa98a8fb39b564de258e68c2/modin/core/storage_formats/pandas/aggregations.py#L402-L421)

In order to make the reduction phase simpler, we can treat all data as having nans. Have you tried this option? There is a feeling that it should not be much worse in terms of performance, but the code will be much simpler.

Reply:
Or complete the vectors that `_compute_non_nan_aggs` function returns to matrices (I suppose this can be done, since they must be symmetrical)

Reply:
Just tried your suggestions (futher referred as 'new logic') by making a square matrix for each aggregation at the Map phase and then handling the Reduction stage as it's a NaN case:

<details><summary>The diffs</summary>

```diff
diff --git a/modin/core/storage_formats/pandas/aggregations.py b/modin/core/storage_formats/pandas/aggregations.py
index e624235a..0b6a0108 100644
--- a/modin/core/storage_formats/pandas/aggregations.py
+++ b/modin/core/storage_formats/pandas/aggregations.py
@@ -215,12 +215,17 @@ class _CorrCovKernels:
             A tuple storing DataFrames where each of them holds the result for
             one of the described aggregations.
         """
+        def make_matrix(sr : pandas.Series) -> pandas.DataFrame:
+            frame = sr.to_frame().iloc[:, [0] * len(sr)]
+            frame.columns = sr.index
+            return frame
+
         sums = df.sum().rename(MODIN_UNNAMED_SERIES_LABEL)
         sums_of_squares = (df**2).sum().rename(MODIN_UNNAMED_SERIES_LABEL)
         count = pandas.Series(
             np.repeat(len(df), len(df.columns)), index=df.columns, copy=False
         ).rename(MODIN_UNNAMED_SERIES_LABEL)
-        return sums, sums_of_squares, count
+        return make_matrix(sums), make_matrix(sums_of_squares), make_matrix(count)

     @staticmethod
     def _compute_nan_aggs(
@@ -329,7 +334,7 @@ class _CorrCovKernels:
         # ...
         # So to get the total result we have to group on the index and sum the values
         total_agg = df.groupby(level=0).sum()
-        total_agg = cls._maybe_combine_nan_and_non_nan_aggs(total_agg)
+        # total_agg = cls._maybe_combine_nan_and_non_nan_aggs(total_agg)

         sum_of_pairwise_mul = total_agg["mul"]
         sums = total_agg["sum"]
@@ -340,25 +345,25 @@ class _CorrCovKernels:
         # If there were NaNs in the original dataframe then have computed a matrix
         # of sums/square sums/counts at the Map phase, meaning that we now have multiple
         # columns in `sums`.
-        has_nans = len(sums.columns) > 1
-        if not has_nans:
-            # 'count' is the same for all columns in a non-NaN case, so converting
-            # it to scalar for faster binary operations
-            count = count.iloc[0, 0]
-            if count < min_periods:
-                # Fast-path for too small data
-                return pandas.DataFrame(index=cols, columns=cols, dtype="float")
-
-            # Converting frame to a Series for more convenient handling
-            sums = sums.squeeze(axis=1)
-            sums_of_squares = sums_of_squares.squeeze(axis=1)
+        # has_nans = len(sums.columns) > 1
+        # if not has_nans:
+        #     # 'count' is the same for all columns in a non-NaN case, so converting
+        #     # it to scalar for faster binary operations
+        #     count = count.iloc[0, 0]
+        #     if count < min_periods:
+        #         # Fast-path for too small data
+        #         return pandas.DataFrame(index=cols, columns=cols, dtype="float")
+
+        #     # Converting frame to a Series for more convenient handling
+        #     sums = sums.squeeze(axis=1)
+        #     sums_of_squares = sums_of_squares.squeeze(axis=1)

         means = sums / count
         std = np.sqrt(sums_of_squares - 2 * means * sums + count * (means**2))

         # The 'is_nans' condition was moved out of the loop, so the loops themselves
         # work faster as not being slowed by extra conditions in them
-        if has_nans:
+        if True:
             return cls._build_corr_table_nan(
                 sum_of_pairwise_mul, means, sums, count, std, cols, min_periods
             )

```

</details>

And here are the perf results for (modin_cpus=112), as it's seen the changes mostly affect pure int cases:

| | old logic | new logic |
|-|-|-|
| [test case from the issue](https://github.com/modin-project/modin/issues/4560#issuecomment-1151498644) (524_288, 100), <br> pure int dataset | 1.48 | 1.69 |
| [test case from the issue](https://github.com/modin-project/modin/issues/4560#issuecomment-1151498644) (524_288, 100), <br> two NaN values in the dataset: | 1.53 | 1.73 |
| [test case from the issue](https://github.com/modin-project/modin/issues/4560#issuecomment-1151498644) (524_288, 100), <br> 20% of the dataset values are NaNs: | 2.95 | 3.07 |
| real-world dataset from our internal <br> workload (1_552_210, 214), 77% of the <br> dataset values are NaNs | 30.6 | 30.58 |
| real-world dataset from our internal <br> workload (1_552_210, 214), but also <br> calling `.fillna(method="ffill")` <br> to get rid of NaN values | 4.13 | 5.46 |

The reasons why the new approach is inefficient:
1. The map's output now have much more data thus stressing the object store, it used to return (ncols, ncols + 3) shaped dataframe, and now it's (ncols, ncols * 4) which is 3 times heavier.
2. The number of columns to aggregate in [this groupby at the reduce phase](https://github.com/dchigarev/modin/blob/b21fb92e5eba3c72b20b604e3aa3e483a0853ea8/modin/core/storage_formats/pandas/aggregations.py#L331) has tripled as well, thus slowing the reduce phase.
3. [`means` and `std`](https://github.com/dchigarev/modin/blob/b21fb92e5eba3c72b20b604e3aa3e483a0853ea8/modin/core/storage_formats/pandas/aggregations.py#L356-L357) computation is now also slower as they now always operating with square matrices rather than 1d vectors
4. The for-loop building the correlation matrix now have [much more index lookups](https://github.com/dchigarev/modin/blob/b21fb92e5eba3c72b20b604e3aa3e483a0853ea8/modin/core/storage_formats/pandas/aggregations.py#L528-L533) that the [non-nan version of the same loop](https://github.com/dchigarev/modin/blob/b21fb92e5eba3c72b20b604e3aa3e483a0853ea8/modin/core/storage_formats/pandas/aggregations.py#L577-L582).

Reply:
@dchigarev thanks for trying, ok.